## Getting MODA Waze Data

MODA gets Waze's crowd sourced traffic data through Waze's Connected Citizens Program. The data feed is in JSON format, and gets updated every 2 minutes. The raw data is stored in Microsoft Azure Storage. 

This notebook shows you how to:
- access the "raw-upload" container in "modawazedata" Azure Storage through their rest API (Azure developer resources here: https://docs.microsoft.com/en-us/rest/api/azure/)
- extract all the filenames of the raw feed in the container
- grab the raw data and save it to a file

Contact Mitsue Iwata, miwata@analytics.nyc.gov, with any questions
Sept 2018

In [2]:
# We have been storing Waze data since June 22nd, 2018. 
# To get an approximation of the number of records stored, we multiply 
# number of weeks * number of days/week * number of hours/day * number of feeds/hour
12 * 7 * 24 * 30

60480

There are roughly 60,480 records from Waze. 

Note, this is a rough approximation. There have been a few short timeframes when the data was not collected 
because of server issues or other hiccups.

## Accessing Azure Storage through REST API

We can call the API to get a list of filenames in a container. However, this only returns 5000 records at a time. Since there is no way to count the number of records, or "blobs", in an Azure container, we will use the approximation from above to loop through our requests in increments of 5000, and use the returned value for "nextmarker" to return the next set of results.


https://docs.microsoft.com/en-us/rest/api/storageservices/list-blobs

### MODA Azure Storage name is: modawazedata
### Container name is: raw-upload



In [3]:
import requests
from bs4 import BeautifulSoup
import lxml
import json
from transform_waze_raw_functions import *

In [4]:
##helper functions
MAXRES_STR = "&maxresults="
MARKER_STR = "&marker="

URL = "https://modawazedata.blob.core.windows.net/raw-upload?restype=container&comp=list&timeout=60"

#build url with max results and next marker
def build_string(max_res, next_marker, url=URL, maxres_str=MAXRES_STR, marker_str=MARKER_STR):
    max_str = maxres_str + str(max_res)
    marker = marker_str + next_marker
    return url + max_str + marker

#make request and soupify
def get_request(url):
    r = requests.get(url)
    c = r.content
    soup = BeautifulSoup(c,'lxml')
    return soup

#generate list of blob names
def make_list_of_blobnames(blob_vals):
    blob_names = []
    for n in blob_vals:
        blob_names.append(n.text)
    return blob_names

#get request and return list of filenames and next marker
def get_filenames_and_marker(soup):
    soup_name = soup.find_all('name')
    
    #generate list of blobnames
    filename_list = make_list_of_blobnames(soup_name)
    
    #get next marker
    marker = soup.find_all('nextmarker')[0].text
    
    return filename_list, marker

#use recursive method with max_results to get all filenames in container 
def get_next_marker_vals(rec_cnt, marker, url, all_blobs):
    if rec_cnt == 0:
        return all_blobs
    else:
        new_url = build_string(rec_cnt, marker, url=url, maxres_str=MAXRES_STR, marker_str=MARKER_STR)
        #print("new URL in recursion ", new_url)
        soup = get_request(new_url)
        blob_list, marker = get_filenames_and_marker(soup)
        all_blobs += blob_list
        print('number of blob filenames ', len(all_blobs))
        rec_cnt = rec_cnt-5000
        print('record count ', rec_cnt)
        
        return get_next_marker_vals(rec_cnt, marker, url, all_blobs)

In [5]:
NUM_RECORDS = 70000 #increments of 5000
MAXRES_STR = "&maxresults="
MARKER_STR = "&marker="

url = "https://modawazedata.blob.core.windows.net/raw-upload?restype=container&comp=list&timeout=60"

#create list object for all blobs names
all_blobs = []

#make first call to url and get the nextmarker value for subsequent calls
s = get_request(url)
blob_list, mark = get_filenames_and_marker(s)
all_blobs += blob_list # add blob names to list

blobnames = get_next_marker_vals(NUM_RECORDS, mark, url, all_blobs)

number of blob filenames  10000
record count  65000
number of blob filenames  15000
record count  60000
number of blob filenames  20000
record count  55000
number of blob filenames  25000
record count  50000
number of blob filenames  30000
record count  45000
number of blob filenames  35000
record count  40000
number of blob filenames  40000
record count  35000
number of blob filenames  45000
record count  30000
number of blob filenames  50000
record count  25000
number of blob filenames  55000
record count  20000
number of blob filenames  60000
record count  15000
number of blob filenames  65000
record count  10000
number of blob filenames  70000
record count  5000
number of blob filenames  75000
record count  0


In [6]:
print(len(blobnames))
print(len(set(blobnames)))

75000
75000


This code isn't perfect - there looks to be some redundancy, but we have about 60K filenames!

In [7]:
blobnames[:10]

['wazeprocessorraw_2018-06-22T20:04:00.3740299Z.json',
 'wazeprocessorraw_2018-06-22T20:06:00.7231344Z.json',
 'wazeprocessorraw_2018-06-22T20:08:00.9739038Z.json',
 'wazeprocessorraw_2018-06-22T20:10:00.6637321Z.json',
 'wazeprocessorraw_2018-06-22T20:12:00.2506104Z.json',
 'wazeprocessorraw_2018-06-22T20:14:00.7759341Z.json',
 'wazeprocessorraw_2018-06-22T20:16:00.2133851Z.json',
 'wazeprocessorraw_2018-06-22T20:18:00.4600203Z.json',
 'wazeprocessorraw_2018-06-22T20:20:00.7461843Z.json',
 'wazeprocessorraw_2018-06-22T20:22:01.1014014Z.json']

## Raw File URL

Now we have to build the url to grab the raw file, then we can write it to a local file.

In [8]:
base_url = "https://modawazedata.blob.core.windows.net/raw-upload/"
blobnames = list(set(blobnames))

file_url_list = []
for b in blobnames:
    file_url_list.append(base_url + b)

file_url_list[:10]

['https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-09-29T20:38:00.3502926Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-07-26T10:48:00.0634092Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-09-22T21:50:00.6489454Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-07-12T00:08:00.1862321Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-08-05T08:56:01.0974502Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-08-23T20:24:00.6667180Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-08-24T02:52:00.5646405Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-07-12T23:26:01.0355045Z.json',
 'https://modawazedata.blob.core.windows.net/raw-upload/wazeprocessorraw_2018-08-11T14:08:01.0583745Z.json',
 'https://modawazed

In [9]:
L1 = []
L2 = []
L3 = []
import numpy as np
for i in np.arange(100):
    req = requests.get(file_url_list[i])
    data = req.json()
    with open('data.json','w') as f:
        json.dump(data,f)
    if 'alerts' in data.keys():
        al = pd.DataFrame(data['alerts'])
        L1.append(al)
    if 'jams' in data.keys():
        jm = pd.DataFrame(data['jams'])
        L2.append(jm)
    if 'irregularities' in data.keys():
        irr = pd.DataFrame(data['irregularities'])
        L3.append(irr)
        
al = pd.concat(L1)
jm = pd.concat(L2)
irr = pd.concat(L3)

al = transform_alerts(al)
jm = transform_jams(jm)
irr = transform_irreg(irr)

al.reset_index(drop=True)
jm.reset_index(drop=True)
irr.reset_index(drop=True)

/Users/Louis/Documents/4A/3_Semester/Capstone/Sandbox/capstone_env/lib/python3.6/site-packages/ipykernel_launcher.py:21: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=True'.

To retain the current behavior and silence the warning, pass sort=False

/Users/Louis/Documents/4A/3_Semester/Capstone/Sandbox/capstone_env/lib/python3.6/site-packages/ipykernel_launcher.py:22: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=True'.

To retain the current behavior and silence the warning, pass sort=False



in jams



/Users/Louis/Documents/4A/3_Semester/Capstone/Sandbox/capstone_env/lib/python3.6/site-packages/pandas/core/frame.py:3778: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


Any questions or suggestions, email, miwata@analytics.nyc.gov

In [10]:
'''
To do:
- How to combine every records with stored filename
- How to add new records from the feed as files are being ETL'd
- How to append new records that come in
'''

"\nTo do:\n- How to combine every records with stored filename\n- How to add new records from the feed as files are being ETL'd\n- How to append new records that come in\n"

In [18]:
al.reset_index(drop=True)

uuid           city state country  \
0      cc657b5d-f45b-3dae-aef4-052c7eb11bcc     Woodbridge    NJ      US   
1      68c3c7fe-e720-32ec-98e8-4d8252e5bc2a     Woodbridge    NJ      US   
2      45beaa65-7059-3083-a588-f9396881ee4b  Staten Island    NY      US   
3      f3eec89f-532e-3e89-92fd-9a39249af58f  Staten Island    NY      US   
4      89fa63e5-9fa9-3b5d-9d6a-d5e359bd68de       Carteret    NJ      US   
5      c66ce900-afb6-339b-8575-73029d420499       Carteret    NJ      US   
6      4843fb56-5083-3f7a-9ed3-41a09a09eb70  Staten Island    NY      US   
7      8e4cc133-fa51-3cb5-8823-1662835d2978      Elizabeth    NJ      US   
8      512e9cd3-1564-3fdd-936d-84e0ed1cd219      Elizabeth    NJ      US   
9      32d535c5-4015-3ec2-b8e9-3770525c2540      Elizabeth    NJ      US   
10     d878271a-587f-307a-bd5a-a85be1184c46      Elizabeth    NJ      US   
11     2d32afd9-0436-3169-9c3c-8d55da7e41c5      Elizabeth    NJ      US   
12     b98a26ba-8b69-3073-86db-eccae4d0435f      Elizabeth    NJ      US   
13     587f0b37-cc35-37f2-8ff7-5c807760c26c      Elizabeth    NJ      US   
14     0c9b4167-67c8-356d-b8a1-6e97c3fad59e         Newark    NJ      US   
15     a8635d97-6a3e-3459-a4ea-26ba661d1e4a         Newark    NJ      US   
16     b1a4b7be-25e7-3d1e-b6b3-b5f71764c374         Newark    NJ      US   
17     6dc4e19b-de65-3f5e-b8ac-ad2736fe11df         Newark    NJ      US   
18     480bf174-d3da-3d9e-b045-f198b5e7408a         Newark    NJ      US   
19     c1a116cf-3f22-3ced-9626-1ca9047a305a         Newark    NJ      US   
20     743485cc-5cb1-3b4d-bc19-c1dc2343e3e0         Newark    NJ      US   
21     eee3acf6-2441-3f13-849b-ba5369c5508f         Newark    NJ      US   
22     da4d7e84-2757-328f-9d80-fe47af8d734c         Newark    NJ      US   
23     a262c89b-31a7-3165-b44f-11f907368009         Newark    NJ      US   
24     be872971-571b-33cc-8227-d6dcc994a6ba         Newark    NJ      US   
25     c5f34b50-538b-3daa-ac65-c3183941bf50         Newark    NJ      US   
26     ce82627e-b23d-34e2-a18b-15b67eee9d03         Newark    NJ      US   
27     4df36a24-4bc9-39fd-9c8e-ecc5aa3d86de         Newark    NJ      US   
28     58778b46-e404-341e-b781-4e43e9df1135         Newark    NJ      US   
29     e5469a7d-e511-338e-94b0-4a1acb86e8d5       Brooklyn    NY      US   
...                                     ...            ...   ...     ...   
58393  6bb88077-d05f-32fd-bf55-5d7a286562ed          Bronx    NY      US   
58394  c874920e-e0ec-3d92-877b-d67a01148b86          Bronx    NY      US   
58395  8a11d191-497e-361b-8545-e4c95040a3a8          Bronx    NY      US   
58396  16be0ffd-b250-3dd1-a7bc-dbfa4230feb4         Leonia    NJ      US   
58397  5ca9e686-ec1a-3cc5-b524-e4c8dc7e5f6b          Bronx    NY      US   
58398  593cf040-fb9f-3c8e-8eb2-35f296ff8111          Bronx    NY      US   
58399  0eea0645-1c8a-322f-bc4b-9ff9baaf29e5         Leonia    NJ      US   
58400  ca459cea-4ff8-34af-b16c-da95a85eccfb          Bronx    NY      US   
58401  3672affa-4614-3c98-8ba6-410557ec2150          Bronx    NY      US   
58402  80e20b4a-ca57-31f7-b38c-6a32ce2057fd          Bronx    NY      US   
58403  cc750ed3-749e-307f-9a61-bb9ff3975c9e          Bronx    NY      US   
58404  35750f66-ca2a-3b5f-966d-4d64023e6d4e         Queens    NY      US   
58405  b7c536bd-cc3b-3ce1-abd6-251efd89a24b         Queens    NY      US   
58406  f9e4faac-ff11-3e3b-9c94-91038b8fc83d         Queens    NY      US   
58407  4f315af2-ce6a-343f-93af-b2c2fb6d4835         Queens    NY      US   
58408  e82016b4-94c1-31b0-b7cb-4f6c2694e459         Queens    NY      US   
58409  f450b28a-1308-3e75-bbb4-bacd080dc0c9         Queens    NY      US   
58410  9eee0416-6d58-3103-90d7-4a0efc6d8407         Queens    NY      US   
58411  7489c958-f131-3fe0-acb4-fdf6a9a02d25          Bronx    NY      US   
58412  bc97b590-2109-37dd-9dbe-b32149b5aa34          Bronx    NY      US   
58413  5f027898-8f27-3ead-b8c9-d914a7b36af4          Bronx    NY      US   
58414  cc1f